milli/src/search/query_tree.rs

```diff
@@ -1310,4 +1384,60 @@ mod test {
             Operation::Query(Query { prefix: true, kind: QueryKind::Exact { .. } })
         ));
     }
+
+    #[global_allocator]
+    static ALLOC: CountingAlloc =
+        CountingAlloc { resident: AtomicI64::new(0), allocated: AtomicI64::new(0) };
+
+    pub struct CountingAlloc {
+        pub resident: AtomicI64,
+        pub allocated: AtomicI64,
+    }
+    unsafe impl GlobalAlloc for CountingAlloc {
+        unsafe fn alloc(&self, layout: std::alloc::Layout) -> *mut u8 {
+            self.allocated.fetch_add(layout.size() as i64, atomic::Ordering::SeqCst);
+            self.resident.fetch_add(layout.size() as i64, atomic::Ordering::SeqCst);
+
+            System.alloc(layout)
+        }
+
+        unsafe fn dealloc(&self, ptr: *mut u8, layout: std::alloc::Layout) {
+            self.resident.fetch_sub(layout.size() as i64, atomic::Ordering::SeqCst);
+            System.dealloc(ptr, layout)
+        }
+    }
+
+    #[test]
+    fn memory_usage_of_ten_word_query() {
+        let resident_before = ALLOC.resident.load(atomic::Ordering::SeqCst);
+        let allocated_before = ALLOC.allocated.load(atomic::Ordering::SeqCst);
+
+        let index = TempIndex::new();
+        let rtxn = index.read_txn().unwrap();
+        let query = "a beautiful summer house by the beach overlooking what seems";
+        let mut builder = QueryTreeBuilder::new(&rtxn, &index).unwrap();
+        builder.words_limit(10);
+        let x = builder.build(query.tokenize()).unwrap().unwrap();
+        let resident_after = ALLOC.resident.load(atomic::Ordering::SeqCst);
+        let allocated_after = ALLOC.allocated.load(atomic::Ordering::SeqCst);
+
+        // Weak check on the memory usage
+        // Don't keep more than 5MB. (Arguably 5MB is already too high)
+        assert!(resident_after - resident_before < 5_000_000);
+        // Don't allocate more than 10MB.
+        assert!(allocated_after - allocated_before < 10_000_000);
+
+        // Use these snapshots to measure the exact memory usage.
+        // The values below were correct at the time I wrote them.
+        // insta::assert_snapshot!(format!("{}", resident_after - resident_before), @"4486950");
+        // insta::assert_snapshot!(format!("{}", allocated_after - allocated_before), @"7107502");
+
+        // Note, with the matching word cache deactivated, the memory usage was:
+        // insta::assert_snapshot!(format!("{}", resident_after - resident_before), @"91248697");
+        // insta::assert_snapshot!(format!("{}", allocated_after - allocated_before), @"125697588");
+        // or about 20x more resident memory (90MB vs 4.5MB)
```

Comment:
This is the where the impact of this PR is measured.

Note that since I had to create a `global_allocator` to measure it, *every other test* of `milli` now also uses this `CountingAlloc`. What are the effects of this? Do we like it, dislike it?

On one hand, it will slow down other tests. On the other hand, we can now measure the memory usage of any operation quite easily.

Reply:
UPDATE: since `cargo_test` runs multiple tests in parallel, this test does not work reliably. I have now commented it out entirely (including the `CountingAlloc`).